tests/difference/core/quint_model/ccv_statemachine.qnt

```diff
@@ -0,0 +1,496 @@
+    action DeliverVscMaturedPacket(sender: Chain): bool =
+        val resultAndTimeout = deliverPacketToProvider(currentState, sender)
+        val result = resultAndTimeout._1
+        all {
+            result.hasError() == false,
+            currentState' = result.newState,
+            trace' = trace.append(emptyAction.with("kind", "DeliverVscMaturedPacket").with("consumerChain", sender))
+        }
+
+    action EndAndBeginBlockForProvider(
+        timeAdvancement: Time,
+        consumersToStart: Set[Chain],
+        consumersToStop: Set[Chain]): bool = 
+        val result = endAndBeginBlockForProvider(currentState, timeAdvancement, consumersToStart, consumersToStop)
+        all {
+            result.hasError() == false,
+            currentState' = result.newState,
+            trace' = trace.append(emptyAction.with("kind", "EndAndBeginBlockForProvider").with("timeAdvancement", timeAdvancement).with("consumersToStart", consumersToStart).with("consumersToStop", consumersToStop))
+        }
+
+    action EndAndBeginBlockForConsumer(
+        chain: Chain,
+        timeAdvancement: Time): bool = 
+        val result = endAndBeginBlockForConsumer(currentState, chain, timeAdvancement)
+        all {
+            result.hasError() == false,
+            currentState' = result.newState,
+            trace' = trace.append(emptyAction.with("kind", "EndAndBeginBlockForConsumer").with("consumerChain", chain).with("timeAdvancement", timeAdvancement))
+        }
+
+    // ==================
+    // INVARIANT CHECKS
+    // ==================
+
+
+    // Every validator set on any consumer chain MUST either be or have been 
+    // a validator set on the provider chain.
+    val ValidatorSetHasExistedInv =
+        runningConsumers.forall(chain =>
+            currentState.consumerStates.get(chain).chainState.votingPowerHistory.toSet().forall(
+                validatorSet => providerValidatorHistory.toSet().contains(validatorSet)
+            )
+        )
+
+    // Any update in the power of a validator on the provider
+    // MUST be present in a ValidatorSetChangePacket that is sent to all registered consumer chains
+    val ValUpdatePrecondition = trace[trace.length()-1].kind == "EndAndBeginBlockForProvider"
+    val ValidatorUpdatesArePropagatedInv = 
+        // when the provider has just entered a validator set into a block...
+        if (ValUpdatePrecondition and currentState.providerState.providerValidatorSetChangedInThisBlock) {
+            val providerValSetInCurBlock = providerValidatorHistory.head()
+            // ... for each consumer that is running then ...
+            runningConsumers.forall(
+                // ...the validator set is in a sent packet
+                consumer => currentState.providerState.sentVscPacketsToConsumer.get(consumer).toSet().exists(
+                    packet => packet.validatorSet == providerValSetInCurBlock
+                )
+            )
+        } else {
+            true
+        }
```

Comment:
I am just curious, why would not you use `A implies B` here? Is it because it would be harder to explain?

Reply:
I just missed it! I changed it to an implication now